### DISCUSSION TITLE
Add method to iterate over models with API keys

### DISCUSSION THREAD

**Discussion by vincentl** [finance-quote/finance-quote#255] [Ideas]
If F::Q provides a mechanism to enumerate the source modules that need an API key, GnuCash can use that method to discover which API keys to advertise to users and, once set by the user, GnuCash can just pass the API key to the Quote.pm new method.

### Preliminary Idea
Up to this point, we have only encountered sites that require a single API key.  That might change in the future, but from GnuCash's perspective, it is just a string that needs to be passed from user to F::Q.  So if we need multiple fields, the string could be a json string with keys and values.

Most modules don't need an API key, so one option is to treat it like https://github.com/finance-quote/finance-quote/blob/501e2deb2c893d479ec8b2ed913f36f39392c4cf/lib/Finance/Quote.pm#L247 and allows modules to optionally define an `api_key_name` method that returns a descriptive string (like "AlphaVantageAPIKey").  

Then create a new method on quoters `$quote->modules_with_api_keys` that returns a hash of `{ module_name => api_key_name_value }`

### Next steps
Need to validate this idea matches what GnuCash can easily use and reach concuss on the F::Q implementation.  Then I am happy to open a merge request with the feature.

**Comment by bpschuck** [comment]
I'm in the in the "if it ain't broke, don't fix it" camp. While having a 
plan for multiple API keys is good, I'd rather not make any changes 
until such time that there are multiple sources requiring API keys.

**Comment by vincentl** [comment]
We have 4 modules (AlphaVantage.pm, IEXCloud.pm, Fixer.pm, OpenExchange.pm) that use API keys.  Two are for securities, two are for currencies.  I don't think there is a way (right now) for GnuCash users to specify which order to try the currency_rate_modules, but at least for the securities we do already have the need for multiple API keys.

I use GnuCash on macOS and right now I need to remember to update a file inside the application bundle to set API keys (I use both AlphaVantage (slower, more securities) and IEXCloud (faster, fewer securities)).  This is not the most user friendly approach.

I started a thread on the gnucash-user email thread because I'm having issues getting F::Q to work on the new macOS Ventura and John Ralls mentioned that having this proposed feature from F::Q would enable GnuCash 5 to allow users to specify F::Q API keys as a GnuCash preference.

I think this proposed change would introduce a new Quote.pm function and would require the 4 modules with API keys to add a method.  The existing environment variable source for API keys can remaining in place (AlphaVantage and IEXCloud first check if Quote->new call has module specific settings and then falls back to the environment variable).

**Comment by bpschuck** [comment]
I forgot about those, as they are likely not used often. Do not forget 
that although GnuCash may the primary use case of F::Q, it is used 
outside of GnuCash. So one should tread lightly when making changes 
specifically for GnuCash. I believe more investigation is require to 
explore other ways to satisfy the issue. Requiring an additional method 
to be included in modules with API keys would be more work for anyone 
contributing a new module and would need to be clearly documented. Seems 
like adding complication just because users forgot to set an environment 
variable. I prefer to follow the KISS protocol.

I have been buried in personal matters relating to a death in a family 
and damages resulting from Hurricane Ian.

**Comment by vincentl** [comment]
@bpschuck - I'm sorry to hear that you are dealing with so much right now and wish you the best in recovering from the damages.

There is no rush to make any decisions and F::Q is definitely much lower in priority than you and your family's well being.

**Comment by bpschuck** [comment]
This discussion should be moved to "Discussions". I'm going to close it here.
BTW - Modules should always support environment variables.

**Reply by bpschuck** [reply]
One thing to remain cognizant of is that the number of users reporting problems with respect to API keys does not represent the majority. Much more probable is that for every user reporting a problem there is likely dozens, hundreds, or even thousands of users who don't have an issue.

Also note that in GnuCash 5.x, the developers will likely make use the `errormsg` label/hash-key to better indicate why F::Q may have failed.

**Reply by jralls** [reply]
This originated with [my post to gnucash-user](https://lists.gnucash.org/pipermail/gnucash-user/2022-December/104063) where I asked for "a Quote.pm function to enumerate the source modules that need an API key".  Adding that shouldn't be hard and shouldn't negatively affect any other F::Q user; in fact I'd think that it would help any app that wants to move away from the legacy API. 

The Quote.pm approach envisioned would be a function `get_sources_with_api_key()` or perhaps more generically `get_sources_requiring_parameter(parameter_name)` and returning a list reference containing the sources that require the given parameter. 

From the other direction the source modules might offer a `get_parameters()` method returning a list of the named parameters needed to operate the module, of which I think `API_KEY` is the only one at present, or `requires_parameter(parameter_name)` returning a bool indicating whether the source module requires the named parameter. @bpschuck given your allergy to coding for needs that aren't yet apparent that could be `requires_api_key()` instead.

Along that line it would also be nice to have `get_description()` and `is_multi()` methods that could be used to populate a selection list. Lacking that GnuCash has to maintain a hard-coded [array of module names and descriptions](https://github.com/Gnucash/gnucash/blob/maint/libgnucash/engine/gnc-commodity.c#L177). We've done a remarkably poor job of keeping that in sync with the sources actually available.

Obviously Quote.pm could wrap the source module methods to return list methods `get_sources_with_api_key()`, `get_single_sources` and `get_multi_sources()` that return listrefs and an enhanced `get_sources` that returns a hashref with the source name as a key with a hash having keys `description`, `single/multi`, and `needs_api_key`. Or not and let the calling code iterate over the list from `get_sources()`.

**Reply by bpschuck** [reply]
John,

Thank you for the well laid out response.

Reading this over I did think of a possible "_gotcha_". For the failover methods (ex: usa, nasdaq, nyse, europe, canada to name a few), how would GnuCash know which API key to supply? For example  both AlphaVantage and IEXCloud may be selected for usa, nyse, or nasdaq methods and GnuCash would not know which one returned the data? Both of those modules use API Keys. Something to think about.

The Quote.pm approach is more appealing to me, as it would not create the need to modify existing modules.

It's not I'm allergic to changes, I am overly cautious as over my career I've seen too many "enhancements" rushed into production that have caused issues.

**Reply by jralls** [reply]
Bruce,

That's not really a problem. The problem is that I confused `sources`, which maps to `methods`, with `modules`. It's `modules` that may take API keys, so the Query.pm function needs to be `get_modules_with_api_key()` instead of `get_sources_with_api_key()`. 

So the flow is:
```
     my $quoter = Finance::Quote->new();
     my @needs_keys = $quoter->get_modules_with_api_key();
     my %key_hash = get_user_api_keys(@needs_keys);
     $quoter = Finance::Quote->new(%key_hash);
     $quoter->fetch($method, @stocks);
```

Implementing `get_modules_with_api_key()` without modifying the modules means hard-coding the modules that need the keys in Quote.pm instead of asking each module whether it does.  The latter is less redundant and so more reliable.

This could be taken a step further with using [Module::Pluggable](https://metacpan.org/pod/Module::Pluggable) to load all of the modules in Quote/ instead of enumerating them Quote.pm. 

> It's not I'm allergic to changes…

I said "allergy to coding for needs that aren't yet apparent" and I was referring to "While having a plan for multiple API keys is good, I'd rather not make any changes until such time that there are multiple sources requiring API keys." in your first comment. There are at present no modules that use parameters other than `API_KEY` so I suggested that you'd prefer `get_needs_api_key()` to `get_needs_parameter('API_KEY')`.

**Reply by vincentl** [reply]
In the last 1-2 years, we introduced multiple currency sources, some of which also require an API key.  A quoter can be constructed with a preferred order of currency modules:

```perl
  my $q = Finance::Quote->new('currency_rates' => {order => ['ECB', 'Fixer'],
                                                   fixer => {API_KEY => $ENV{FIXER_API_KEY}}});
```

There are at least two ways to get security methods, `get_methods()` is a "class" method and doesn't require a quoter object.  So we could introduce one class method `get_all_sources()` that returns a deep json-like data structure:

```perl
{ 'currency_rate_modules' =>
   [ {'name' => 'ECB', 'parameters' => []},
      {'name' => 'Fixer', 'parameters' => ['FIXER_API_KEY']}],
  'modules' => 
   [ {'name' => 'AlphaVantage', 'parameters' => ['ALPHAVANTAGE_API_KEY']},
     {...}]
}
```

Not sure I like this and not sure what is most convenient for client code.

This would require introducing one new Quote.pm function that would iterate over the `@CURRENCY_RATES_MODULES` and `@MODULES` arrays and for each do something similar to `$modpath->can('get_public_parameters') || []` to populate the parameter array.

**Comment by bpschuck** [comment]
@jralls If the user selects the method `usa` in the security editor for quote sources, how would GnuCash know which API  Key (if any) to supply? The `usa` method  could be AlphaVantage, IEXCloud, or any other module that presents the `usa` method (I'm thinking YahooJSON should return more methods than just `yahoo_json`). 

Looking at the master branch of the GnuCash repo, it appears that the 5.x release will use a different mechanism other than gnc-fq-helper to fetch quotes. Is that correct John?

@vincentl - I think John would like to see something along the line of `get_all_sources()`. To help address my question about the failback methods like `usa` it would be handy to return the different methods presented as part of the JSON if we go that route. Another idea would be to have Quote.pm function that returns a similar JSON structure containing only those modules for a given method. `get_modules_containing('nyse')` returns data pertaining to only those modules that present the `nyse` method. Something like that, thinking out loud at this point.

**Comment by jralls** [comment]
@bpschuck 
>@jralls If the user selects the method usa in the security editor for quote sources, how would GnuCash know which API Key (if any) to supply? The usa method could be AlphaVantage, IEXCloud, or any other module that presents the usa method (I'm thinking YahooJSON should return more methods than just yahoo_json).

IIUC GnuCash needs to know what *modules* need API keys and to pass those keys that it has to `$quoter = F::Q->new()`.  `$quoter->fetch()` takes a *method* as an arg; if that method winds up trying to use a module that needs an API key that wasn't supplied  when instantiating `$quoter` I'd expect it to complain about it in the error stream and failover to the next module for that method.

> Looking at the master branch of the GnuCash repo, it appears that the 5.x release will use a different mechanism other than gnc-fq-helper to fetch quotes. Is that correct John?


Yes. `gnc-fq-check`, `gnc-fq-dump`, and `gnc-fq-helper` are replaced by a very thin [`gnc-fq-wrapper`](https://github.com/Gnucash/gnucash/blob/master/libgnucash/quotes/finance-quote-wrapper.in) that forwards queries to F::Q and passes back the results as a pair of JSON streams, one for data and one for errors. Those streams are read and interpreted in [`gnc-quotes.cpp`](https://github.com/Gnucash/gnucash/blob/master/libgnucash/app-utils/gnc-quotes.cpp) where there's access to GnuCash's logging facility and where the GUI can easily see the outcome and report it to the user.

**Comment by jralls** [comment]
>   I think John would like to see something along the line of get_all_sources(). To help address my question about the failback methods like usa it would be handy to return the different methods presented as part of the JSON if we go that route. Another idea would be to have Quote.pm function that returns a similar JSON structure containing only those modules for a given method. get_modules_containing('nyse') returns data pertaining to only those modules that present the nyse method. 

I guess the objective is to create a list of required modules based on the list of methods for which the user has configured quote retrieval. That's a worthwhile goal, but it's complicated by the fact that multi-source `mmm` might be set up to check methods `aaa`, `bbb`, and `ccc` where `aaa` and `ccc` require passwords. The user wants a price for `xyz` which is usually available from `aaa` and `bbb`.  How would the client (GnuCash) know that it doesn't need to pester the user for an API key for `ccc`?

**Comment by vincentl** [comment]
If there is separation between choosing sources for a security and setting module parameters, then GnuCash would construct the quoter with whatever optional module parameters (and possible currency module preferences) the user has specified in some preference panel.  Separately, the user has selected what source to use for security.  When the quoter is used to fetch, it the source is a single module it will either succeed for fail with a "Need API Key" type error, if the source is multiple modules, then failover will silently skip over modules that wanted an API key but weren't provided.  This could be a problem if that was the only module in the failover sequence that could fetch the quote.   I think right now, the error will come from the final module (not sure, need to double check).

To populate the possible sources for a security, GnuCash is already asking FQ for the list of sources and that can stay as is.  (Or, if a source requires a parameter, maybe that would be indicated somehow in the source list, but the parameters would be set elsewhere).

To populate this new preference panel, GnuCash would just need to know which modules need parameters and the names of those parameters.

Is this making sense?

**Reply by jralls** [reply]
Except for the fail silently part. If a module needs an api key and one isn't provided it should put an error on the error stream including which symbol it was attempted for. Then the client can compare that against the list of symbols for which prices were retrieved and include the missing key as part of the reason that the price wasn't retrieved.

**Comment by vincentl** [comment]
I just opened PR #260 as an experimental implementation of the `get_features()` class method that I think will provide all the information client code such as GnuCash will need.

**Comment by bpschuck** [comment]
I will have to look when I have time. Currently in another round of 
legal/insurance matters consuming all my free time.

I'd also like to see John Ralls or another GnuCash core developer to 
confirm they could and would use it *before* being merged into the 
master branch.

Thanks.

- Bruce S.

**Reply by jralls** [reply]
Unless someone comes forward to write a PR that will have to wait until after we get GnuCash 5.0 released. The target is [March 28](https://wiki.gnucash.org/wiki/Release_Schedule#Goals_for_5.0).

**Comment by bpschuck** [comment]
On Sat, 21 Jan 2023 11:09:02 -0800 Joh Ralls wrote:

> Vincent,

> The list of F::Q sources displayed in the Security Editor dialog is
> partially hard-coded in libgnucash/engine/gnc-commodity.c beginning
> at line 171. Every entry in those two lists represents an F::Q
> method. At startup GC calls F::Q's `$quoter->sources()` and compares
> the returned list against the hard-coded lists, enabling the matching
> entries. Any F::Q methods not in either hard-coded list is added to a
> new dynamic list that's displayed as the "unknown" list--unknown
> because at present GnuCash has no way of knowing whether it's a
> single or a multi-source method--in F::Q terminology whether it has
> fallbacks. Maintaining that list is a pain in the rear and its
> existence creates a coupling between GC and F::Q versions. The same
> applies to modules and API keys and to which module provides which
> methods.

.
.
.

> That's a mid-size project. I don't think it will be possible to have
> it ready for GnuCash 5.0, whose release is only 10 weeks from now
> with feature freeze in only two weeks. It might be possible to get it
> into 5.1 at the end of June but I think 5.2 at the end of September
> is a more realistic goal.

> See https://wiki.gnucash.org/wiki/CodingStandard and
> https://wiki.gnucash.org/wiki/C++ for style guidance.

Gentlemen,

May I suggest this discussion get moved or copied to the discussion list 
on https://github.com/finance-quote/finance-quote, possibly 
https://github.com/finance-quote/finance-quote/discussions/255 or 
perhaps a new one (In addition I think adding 
***@***.*** to the 
Cc list may copy that discussion). In my opinion the GitHub discussion 
feature is maybe a better fit for discussing what possible changes may 
be made to F::Q and underlying modules than the Gnucash-devel mailing 
list. I think following a discussion may be easier than email list 
thread/subject archives.

Thank you John for giving a more detailed description of the GnuCash to 
Finance::Quote interface and explaining how more thought and planning 
would benefit both GnuCash and F::Q.

- Bruce

**Reply by jralls** [reply]
@bpschuck, that email is about *GnuCash* code, as @vincentl is preparing to develop a GnuCash PR. It was a response to [this email](https://lists.gnucash.org/pipermail/gnucash-devel/2023-January/046534.html). An earlier exchange about building GnuCash is in [this thread](https://lists.gnucash.org/pipermail/gnucash-devel/2023-January/046516.html). IMO conversations about how GnuCash code works belong in the GnuCash mailing lists, not on F::Q's GitHub repo where they're somewhat irrelevant.

I think it's more appropriate to link the emails in the mailing list archive so that others can easily get the context rather than to copy them.

**Comment by bpschuck** [comment]
Copied from John Ralls' post to the GnuCash-devel mailing list:

The list of F::Q sources displayed in the Security Editor dialog is partially hard-coded in libgnucash/engine/gnc-commodity.c beginning at line 171. Every entry in those two lists represents an F::Q method. At startup GC calls F::Q's `$quoter->sources()` and compares the returned list against the hard-coded lists, enabling the matching entries. Any F::Q methods not in either hard-coded list is added to a new dynamic list that's displayed as the "unknown" list--unknown because at present GnuCash has no way of knowing whether it's a single or a multi-source method--in F::Q terminology whether it has fallbacks. Maintaining that list is a pain in the rear and its existence creates a coupling between GC and F::Q versions. The same applies to modules and API keys and to which module provides which methods. The perl interface is in libgnucash/quotes/finance-quotes-wrapper.in that's turned into bin/finance-quotes-wrapper.pl at configuration. Any additions or changes there should be limited t
 o converting a JSON stream on stdin to function arguments, calling the F::Q function, and converting the return data into a JSON stream on stdout and errors on stderr. The wrapper is driven by gnucash/price-quotes.cpp that's responsible for presenting an API to the rest of the program. API callsl set arguments and marshal them into the JSON stream sent to finance-quoter-wapper's stdin, invoking the appropriate finance-quote-wrapper function, parsing the returned JSON string into whatever C++ struct the API promises to the rest of GnuCash, and parsing and interpreting the stderr JSON. finance-quote-wrapper functions should be kept as thin as possible because it's perl running in a separate process and must interact purely through the JSON streams. The pair of finance-quote-wrapper and price-quotes.cpp should abstract away as much implementation detail as possible so that coupling between F::Q and the rest of GnuCash is minimal.

I think that the user interface, both for the source selection in the GUI and for retrieving specified quotes from the command line, should continue to focus on methods and treat the methods behind them as much as an implementation detail as possible. To accomplish that GnuCash needs to know when a method is implemented by a module that needs an API key and which key that is. If GC doesn't already have that key stored it needs to prompt for it using a dialog box with a "remember" checkbox in the GUI; there are several ways one could go with a terminal command of which erroring out (`method foo requires a bar API key and you haven't provided one`).

I'd also like to get rid of the hard-coded method lists. That requires being able to query which methods are single-source and which have fallbacks. For the latter it would be nice to have a tooltip listing what single-source methods in what order will be tried to obtain the quote. It would also be nice to flag on the list which sources need which API keys. GC also needs to scan the configured securities for sources to identify any sources that aren't provided by the installed F::Q and present that as an error to the user.

Currency sources are interesting in a different way. Last I looked F::Q had four available sources for looking up currencies defaulting on alphavantage. Using anything other than the F::Q default will require substantially changing the way GC requests currency rates. There are probably users who would like to configure sources per pair of currencies, e.g ECB for EUR<->USD and Alphavantage for USD<->INR. Others might be content with a preference setting for the currency method. Most may not care at all as long as they can get the rates that they want without much fiddling.

That's a mid-size project. I don't think it will be possible to have it ready for GnuCash 5.0, whose release is only 10 weeks from now with feature freeze in only two weeks. It might be possible to get it into 5.1 at the end of June but I think 5.2 at the end of September is a more realistic goal.

See https://wiki.gnucash.org/wiki/CodingStandard and https://wiki.gnucash.org/wiki/C++ for style guidance.

**Comment by vincentl** [comment]
Starting a [WIP:PR](https://github.com/Gnucash/gnucash/pull/1626) for GnuCash to use this feature.

**Comment by bpschuck** [comment]
PR #260 has been merged into master branch.

**Comment by vincentl** [comment]
Closing discussion after PR merge
